cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/Partitioning.java

```diff
@@ -0,0 +1,215 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.dataset.lib;
+
+import co.cask.cdap.api.common.Bytes;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import javax.annotation.Nonnull;
+
+/**
+ * This class describes how a dataset is partitioned, by means of the fields of a partition key and their types.
+ */
+public class Partitioning {
+
+  /**
+   * Describes the type of a partitioning field.
+   */
+  public enum FieldType {
+    STRING {
+      public String parse(String value) {
+        return value;
+      }
+      public <T> byte[] toBytes(T value) {
+        return Bytes.toBytes((String) value);
+      }
+      public int determineLengthInBytes(byte[] bytes, int offset) {
+        for (int i = offset; i < bytes.length; i++) {
+          if (bytes[i] == 0) {
+            return i - offset;
+          }
+        }
+        return bytes.length - offset;
+      }
+      public String fromBytes(byte[] bytes, int offset, int length) {
+        return Bytes.toString(bytes, offset, length);
+      }
+      public boolean validateType(Object value) {
+        return value instanceof String;
+      }
+      public String toHiveType() {
+        return "STRING";
+      }
+    },
+    LONG {
+      public Long parse(String value) {
+        return Long.parseLong(value);
+      }
+      public <T> byte[] toBytes(T value) {
+        // trick to preserve the ordering of the generated byte array
+        return Bytes.toBytes(((Long) value) ^ Long.MIN_VALUE);
+      }
+      public int determineLengthInBytes(byte[] bytes, int offset) {
+        return Bytes.SIZEOF_LONG;
+      }
+      public Long fromBytes(byte[] bytes, int offset, int length) {
+        return Bytes.toLong(bytes, offset, length) ^ Long.MIN_VALUE;
+      }
+      public boolean validateType(Object value) {
+        return value instanceof Long;
+      }
+      public String toHiveType() {
+        return "BIGINT";
+      }
+    },
+    INT {
+      public Integer parse(String value) {
+        return Integer.parseInt(value);
+      }
+      public <T> byte[] toBytes(T value) {
+        // trick to preserve the ordering of the generated byte array
+        return Bytes.toBytes(((Integer) value) ^ Integer.MIN_VALUE);
+      }
+      public int determineLengthInBytes(byte[] bytes, int offset) {
+        return Bytes.SIZEOF_INT;
+      }
+      public Integer fromBytes(byte[] bytes, int offset, int length) {
+        return Bytes.toInt(bytes, offset, length) ^ Integer.MIN_VALUE;
+      }
+      public boolean validateType(Object value) {
+        return value instanceof Integer;
+      }
+      public String toHiveType() {
+        return "INT";
+      }
+    };
+
+    public abstract Comparable parse(String value);
+    public abstract <T> byte[] toBytes(T value);
```

Comment:
no. If FieldType is STRING, then T should be String. I am not super happy with this class, this is where you hit the limits of the Java type system, and generics only get you that far.